File: JS/Tree/938.Range Sum of BST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} low
 * @param {number} high
 * @return {number}
 */
var rangeSumBST = function(root, low, high) {
    let sum = 0;
    const stack = [root];
    while (stack.length) {
        const node = stack.pop();
        if (node) {
            if (node.val >= low && node.val <= high) {
                sum += node.val;
            }
            if (node.val > high) {
                stack.push(node.left);
            } else if (node.val < low) {
                stack.push(node.right);
            } else {
                stack.push(node.left);
                stack.push(node.right)
            }
        }
    }
    return sum;
};

File: Golang/Tree/938.Range-sum-of-BST.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rangeSumBST(root *TreeNode, low int, high int) int {
	sum := 0
	current := root
	for current != nil {
		if current.Left == nil {
			if current.Val >= low && current.Val <= high {
				sum += current.Val
			}
			current = current.Right
		} else {
			prev := current.Left
			for prev.Right != nil && prev.Right != current {
				prev = prev.Right
			}
			if prev.Right == nil {
				prev.Right = current
				current = current.Left
			}
			if prev.Right == current {
				prev.Right = nil
				if current.Val >= low && current.Val <= high {
					sum += current.Val
				}
				current = current.Right
			}
		}
	}
	return sum
}
